### PR TITLE
🐛 amp-list: Don't wait for amp-bind before first mutate

### LINKED ISSUE #11998: [amp-bind] amp-list evaluation does not wait for amp-state's src request to be resolved
Bindings are evaluated when an amp-list template is rendered. However, these bindings don't wait until an amp-state's remote src has been resolved (which they should to ensure an deterministic behavior).

**Example:** 

An amp-state element using a JSON endpoint:
```
<amp-state id="favorite"
  credentials="include"
  src="/favorite-with-count?RANDOM">
</amp-state>
```

The amp-list binds an element to the amp-state:
```
<amp-list ... src="/favorite-with-count?RANDOM">
    <template type="amp-mustache">
        <div class="favorite-count" [text]="favoriteWithCount.count ? favoriteWithCount.count : ''">{{count}}</div>
    </template>
  </amp-list>
```

This results in a race-condition, where the text is either `null` or `123` depending on whether the amp-state has already been resolved or not.

I noticed this while working on this sample: https://ampbyexample.com/advanced/favorite_button/

//cc @choumx @ericlindley-g

### LINKED ISSUE #15311: amp-list: Don't evaluate bindings until first user gesture
https://github.com/ampproject/amphtml/blob/a42a4a8bc73bfedb15937a441fa482503fa2eeda/extensions/amp-list/0.1/amp-list.js#L291-L299

We shouldn't wait for `amp-bind` to be installed and registered as a service before rendering. Binding evaluation should be a no-op before the first user gesture anyways, so we should skip this step until then. 

This means that on the first user gesture, `amp-bind` should toggle some commonly available flag. `amp-list` can read that flag to decide whether to block on binding evaluation or not.

Related: #15272, #11998

### PR DESCRIPTION
**Pull request by choumx**
Fixes #15311 and fixes #11998. Partial for #15272.

Dramatically reduces initial render latency for `<amp-list>` elements on pages that also use `<amp-bind>`. Looks like ~95% improvement by my simulations on `examples/bind/list.amp.html`. 

A picture of me not thinking things through thoroughly enough:

image

/to @aghassemi

### CODE REVIEW

File: build-system/global-configs/canary-config.json [hunk 1]
```diff
@@ -3,6 +3,7 @@
     "amp-date-picker",
     "amp-next-page",
     "ampdoc-shell",
+    "disable-faster-amp-list",
```
Comment (aghassemi) [inline build-system/global-configs/canary-config.json]: lol

File: extensions/amp-list/0.1/amp-list.js [hunk 2]
```diff
@@ -282,18 +290,44 @@ export class AmpList extends AMP.BaseElement {
...
+    // Allow a temporary opt-out to this behavior in case this causes breaking
+    // changes due to state-independent mutations, e.g. "[text]="1+1".
+    if (isExperimentOn(this.win, 'disable-faster-amp-list')) {
+      return Services.bindForDocOrNull(this.element).then(bind => {
+        if (bind) {
+          return this.updateBindingsWith_(bind, elements);
+        }
+      });
+    } else {
+      // Skip if setState() has _not_ been invoked yet.
```
Comment (aghassemi) [inline extensions/amp-list/0.1/amp-list.js]: nit, comment applies to the else case

**Comment by sebastianbenz** [comment]
Won't this break existing implementations? For example, the [favicon sample](https://ampbyexample.com/advanced/favorite_button/) relies on bindings being evaluated when list content is rendered (I thought this was intentional).

**Comment by choumx** [comment]
It shouldn't break use cases that rely on user interaction since it'll still evaluate bindings in list content after the first `AMP.setState`.

**Comment by sebastianbenz** [comment]
In that case it will break the [favourite sample](https://ampbyexample.com/advanced/favorite_button/) as it relies on bindings being evaluated when the list is loaded. The sample should be easy to fix, but we might break other sites.

**Comment by aghassemi** [comment]
@choumx so we can no longer render `amp-list` from `amp-state` for an initial stale render?

**Comment by sebastianbenz** [comment]
`<amp-list src="foo.bar">` with `<amp-state id="foo">` would be great!

**Comment by ericlindley-g** [comment]
Is it possible to understand how many sites could be affected by this change before launching it? (e.g. other sites that have copied the model used in the "favorite" sample)

If I recall correctly, we intentionally decided to block amp-list render on evaluating bind expressions contained in those amp-lists — will this change undo that?

**Pull request merged**